ssl/statem/extensions.c

```diff
@@ -929,9 +929,27 @@ static int final_server_name(SSL *s, unsigned int context, int sent)
         ret = s->session_ctx->ext.servername_cb(s, &altmp,
                                        s->session_ctx->ext.servername_arg);
 
-    if (!sent) {
-        OPENSSL_free(s->session->ext.hostname);
-        s->session->ext.hostname = NULL;
+    /*
+     * For servers, propagate the SNI hostname from the temporary
+     * storage in the SSL to the persistent SSL_SESSION, now that we
+     * know we accepted it.
+     * Clients make this copy when parsing the server's response to
+     * the extension, which is when they find out that the negotiation
+     * was successful.
+     */
+    if (s->server) {
+        if (!sent) {
+            OPENSSL_free(s->ext.hostname);
+            s->ext.hostname = NULL;
```

Comment:
It might be better to set s->ext.hostname to NULL at init_server_name()

Reply:
Ben, Can you please describe a scenario where sent is false, and s->ext.hostname isn't NULL ?
I can think of a scenario that involves the sni callback that sets s->ext.hostname,
but in such case, I don't think that s->ext.hostname should be set to NULL.

Reply:
I believe in a TLS 1.2 resumption case where the client sends SNI and we successfully resume, the server will not have sent SNI in the ServerHello (```sent == 0```), but ```s->ext.hostname``` would have been set in ```tls_parse_ctos_server_name()```, so this free+NULL would have significance.

Reply:
'sent' arg is 'raw_extension_st::present',
meaning that when final_server_name() is called at the server side (s->server is true),
then 'sent' doesn't mean that the server sent this extension.
'sent' means that the client sent this extension, and that the server received this extension.
so if sent is false, then s->ext.hostname is NULL, unless server name callback, or another callback has set it.
so it seems that this null assignment isn't needed.​

Reply:
This analysis looks reasonable within the bounds of a single handshake, but the situation gets more complicated when (pre-TLS 1.3) renegotiation is in play.  A hostname value could have been set by an initial handshake and then omitted from a renegotiation handshake, in which case this clearing code seems like it would still have non-trivial effect.